### ISSUE TITLE
`up` or other commands could fail from partially completed `init`

### ISSUE THREAD

**Issue by timheuer** [Azure/azure-dev#3053]
**Describe the bug**
If you run an `azd init` and it partially fails, the process is incomplete, creating an app environment but not an azure.yaml file.  Thus subsequent operations fail with:

```
Error: no project exists; to create a new project, run `azd init`
```

**To Reproduce**
Create a new .NET Aspire app
In the app host add PostgresContainer (not yet supported)
`azd init`
observe that init will fail saying the resource type is not supported
remove the postgrescontainer from the app host
run `azd up` -- fails
run `azd init` -- says app env already created

**Expected behavior**
Some way to 're-init'?

**Comment by rajeshkamal5050** [comment]
Currently for such failure scenarios/partial state. We have to delete the `.azure` folder and `azure.yaml`. Re-run azd init from the root of your solution as a re-init strategy.

@weikanglim can we handle failure scenarios/state better. So subsequent `azd init` works fine?

**Comment by vhvb1989** [comment]
azd should either:
- write all files at the end of azd init
- have a revert-changes routine to clean whatever is created if there is an error.

The current model is a bug, as it lets the project on invalid state.  This could be an easy fix, @weikanglim ?

**Comment by weikanglim** [comment]
@vhvb1989 A partial completion is not an invalid state. It is only unhandled in the current model. We've brought up reinitialization as an improvement for a while now. Making it idempotent may be worthwhile.

My initial thought is that an undo model that isn't based on atomic rollback transactions, but rather strict "undo-s" will continue to expand as the scope of the logic expands to "what should be undone?"

**Comment by weikanglim** [comment]
Here's my proposal:

The core `azd init` initialize an environment really means currently:

1. Create `.azure/<env>`
2. Make `.azure/config.json` point to it as a default environment

We update the code here: on a rerun, if there is a default environment, and the environment is valid, we can proceed.

In the "extended init flow" we have the following additional env modifications:

1. In template initialization, we allow for template metadata to set environment values/config. We update the code here to perform a diff on what changes would be made, with a confirmation prompt.
2. In Aspire's initialization, we prompt for exposed services. We update the code here to prompt with default values based on the stored exposed services.

Flags for azd init:

If `--environment / -e` is passed, `--subscription` and `--location` can also be passed to create an environment. It is however invalid to use `--subscription` and `--location` to set an existing environment.

**Comment by vhvb1989** [comment]
@weikanglim , I am not a fan of supporting `crashed states` as valid and then making azd to check what's the current state for either continue or start from zero.

To me, all single-commands should be atomic-transactions.  Only composed-commands (like azd up) should allow intermediate states.

This is one of the reasons why azd prefers `fail fast`, over failing on the middle of a process and bringing doubts to customers about what to do next.

My preference would be run any validations we know could fail and fails as soon as possible. Never at the middle or after applying changes locally.

**Comment by weikanglim** [comment]
@vhvb1989 I disagree that this is a "crashed state" because there is no corruption. A user can stumble upon the same scenario by manually creating `azure.yaml`, running `azd env new`. A user could also choose to reinitialize from a template to obtain files and environment variables. This is a perfectly valid state to be in, azd should handle that. azd should support reinitialization as a concept, here we're starting down that path.

> My preference would be run any validations we know could fail and fails as soon as possible. Never at the middle or after applying changes locally.

This sounds nice in theory, but in-practice, running the validation could incur computational costs near the operations being performed. Thus, you are hit with the operational cost twice.
